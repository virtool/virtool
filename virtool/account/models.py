from datetime import datetime

from email_validator import EmailSyntaxError, validate_email
from pydantic import ConstrainedStr, validator

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.enums import AnalysisWorkflow
from virtool_core.models.group import GroupMinimal, Permissions
from virtool_core.models.user import User


def check_email(email: str | None) -> str | None:
    """Checks if the given email is valid."""
    if email is None:
        return None

    try:
        validate_email(email)
    except EmailSyntaxError:
        raise ValueError("The format of the email is invalid")

    return email


class AccountSettings(BaseModel):
    quick_analyze_workflow: AnalysisWorkflow
    show_ids: bool
    show_versions: bool
    skip_quick_analyze_dialog: bool


class ConstrainedEmail(ConstrainedStr):
    strip_whitespace = True


class Account(User):
    settings: AccountSettings
    email: ConstrainedEmail | None

    _email_validation = validator("email", allow_reuse=True)(check_email)


class APIKey(BaseModel):
    id: str
    created_at: datetime
    groups: list[GroupMinimal]
    name: str
    permissions: Permissions
