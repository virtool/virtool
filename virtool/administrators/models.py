from typing import List

from virtool_core.models.basemodel import BaseModel
from virtool_core.models.roles import AdministratorRole
from virtool_core.models.user import UserNested


class AdministratorMinimal(UserNested):
    role: AdministratorRole


class Administrator(AdministratorMinimal):
    available_roles: List[dict]


class AdministratorSearch(BaseModel):
    items: List[AdministratorMinimal]
    available_roles: List[dict]
