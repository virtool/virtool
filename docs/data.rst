Data Layer
==========

.. automodule:: virtool.data.__init__
    :members:

Transforms
----------

.. automodule:: virtool.data.transforms


API
+++

.. autoclass:: virtool.data.transforms.AbstractTransform
    :members:

.. autofunction:: virtool.data.transforms.apply_transforms

Errors
------

.. automodule:: virtool.data.errors
    :members:


