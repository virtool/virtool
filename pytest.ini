[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
