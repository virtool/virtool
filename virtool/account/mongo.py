"""Work with the current user account and its API keys."""

from typing import Any

import virtool.users.utils
import virtool.utils
from virtool.mongo.core import Mongo


def compose_password_update(password: str) -> dict[str, Any]:
    """Compose an update dict for self-changing a users account password.

    This will disable forced reset and won't invalidate current sessions, unlike a
    password change by an administrator.

    :param password: the new password
    :return: a password update

    """
    return {
        "password": virtool.users.utils.hash_password(password),
        "invalidate_sessions": False,
        "last_password_change": virtool.utils.timestamp(),
        "force_reset": False,
    }


async def get_alternate_id(mongo: Mongo, name: str) -> str:
    """Get an alternate id for an API key whose provided `name` is not unique. Appends
    an integer suffix to the end of the `name`.

    :param mongo: the application mongodb client
    :param name: the API key name
    :return: an alternate unique id for the key

    """
    existing_alt_ids = await mongo.keys.distinct("id")

    suffix = 0

    while True:
        candidate = f"{name.lower()}_{suffix}"

        if candidate not in existing_alt_ids:
            return candidate

        suffix += 1
