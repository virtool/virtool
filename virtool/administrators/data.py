from sqlalchemy.ext.asyncio import AsyncEngine
from virtool_core.models.roles import AdministratorRole

from virtool.administrators.actions import get_action_from_name
from virtool.authorization.client import AuthorizationClient
from virtool.authorization.relationships import AdministratorRoleAssignment
from virtool.data.domain import DataLayerDomain
from virtool.mongo.core import Mongo

PROJECTION = [
    "_id",
    "handle",
    "administrator",
    "force_reset",
    "groups",
    "last_password_change",
    "permissions",
    "primary_group",
    "administrator_role",
    "active",
    "b2c",
    "b2c_display_name",
    "b2c_family_name",
    "b2c_given_name",
    "b2c_oid",
]


class AdministratorsData(DataLayerDomain):
    name = "administrators"

    def __init__(
        self, authorization_client: AuthorizationClient, mongo: Mongo, pg: AsyncEngine
    ):
        self._authorization_client = authorization_client
        self._mongo = mongo
        self._pg = pg

    async def run_action(self, name: str):
        """
        Run an action

        Runs an action with the given name.

        :param name: the name of the action to run
        :return: the result of the action
        """
        return await get_action_from_name(name).run(self.data)
