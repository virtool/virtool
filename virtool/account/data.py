from aioredis import Redis
from sqlalchemy.ext.asyncio import AsyncEngine
from virtool_core.models.account import Account
from virtool_core.models.account import AccountSettings, APIKey
from virtool_core.models.session import Session

import virtool.utils
from virtool.account.mongo import (
    compose_password_update,
)
from virtool.account.oas import (
    UpdateSettingsRequest,
    CreateKeysRequest,
    UpdateKeyRequest,
    ResetPasswordRequest,
    CreateLoginRequest,
    UpdateAccountRequest,
)
from virtool.administrators.oas import UpdateUserRequest
from virtool.authorization.client import AuthorizationClient
from virtool.data.domain import DataLayerDomain
from virtool.data.errors import ResourceError, ResourceNotFoundError
from virtool.data.transforms import apply_transforms
from virtool.groups.transforms import AttachGroupsTransform
from virtool.mongo.core import Mongo
from virtool.mongo.utils import get_one_field
from virtool.users.mongo import validate_credentials
from virtool.users.utils import limit_permissions
from virtool.utils import base_processor

PROJECTION = (
    "_id",
    "email",
    "handle",
    "groups",
    "last_password_change",
    "permissions",
    "primary_group",
    "settings",
    "force_reset",
)


class AccountData(DataLayerDomain):
    name = "account"

    def __init__(
        self,
        mongo: Mongo,
        redis: Redis,
        authorization_client: AuthorizationClient,
        pg: AsyncEngine,
    ):
        self._authorization_client = authorization_client
        self._mongo = mongo
        self._redis = redis
        self._pg = pg

    async def get(self, user_id: str) -> Account:
        """
        Get the account for the given ``user_id``.

        :param user_id: the user ID
        :return: the user account
        """
        user = await self.data.users.get(user_id)

        if user:
            return Account(
                **{
                    **user.dict(),
                    "settings": {
                        "quick_analyze_workflow": "nuvs",
                        "show_ids": False,
                        "show_versions": False,
                        "skip_quick_analyze_dialog": False,
                    },
                }
            )

        raise ResourceNotFoundError("User not found")

    async def update(self, user_id: str, data: UpdateAccountRequest) -> Account:
        """
        Update the user account.

        :param user_id: the user ID
        :param data: the update to the account
        :return: the user account
        """
        update = {}

        data_dict = data.dict(exclude_unset=True)

        if "password" in data_dict:
            if not await validate_credentials(
                self._mongo, user_id, data_dict["old_password"] or ""
            ):
                raise ResourceError("Invalid credentials")

            update = compose_password_update(data_dict["password"])

        if "email" in data_dict:
            update["email"] = data_dict["email"]

        if update:
            result = await self._mongo.users.update_one(
                {"_id": user_id}, {"$set": update}
            )

            if result.modified_count == 0:
                raise ResourceNotFoundError("User not found")

        return await self.get(user_id)

    async def get_settings(self, query_field: str, user_id: str) -> AccountSettings:
        """
        Gets account settings.

        :param query_field: the field to get
        :param user_id: the user ID
        :return: the account settings
        """
        settings = await get_one_field(self._mongo.users, query_field, user_id)

        return AccountSettings(**settings)

    async def update_settings(
        self, data: UpdateSettingsRequest, query_field: str, user_id: str
    ) -> AccountSettings:
        """
        Updates account settings.

        :param data: the update to the account settings
        :param query_field: the field to edit
        :param user_id: the user ID
        :return: the account settings
        """
        settings_from_mongo = await get_one_field(
            self._mongo.users, query_field, user_id
        )

        data_dict = data.dict(exclude_unset=True)

        settings = {**settings_from_mongo, **data_dict}

        await self._mongo.users.update_one({"_id": user_id}, {"$set": settings})

        return AccountSettings(**settings)

    async def get_keys(self, user_id: str) -> list[APIKey]:
        """
        Gets API keys associated with the authenticated user account.

        :param user_id: the user ID
        :return: the api keys
        """

        keys = await apply_transforms(
            [
                base_processor(key)
                async for key in self._mongo.keys.find({"user.id": user_id})
            ],
            [
                AttachGroupsTransform(self._pg),
            ],
        )

        return [APIKey(**key) for key in keys]

    async def get_key(self, user_id: str, key_id: str) -> APIKey:
        """
        Get the complete representation of the API key identified by the `key_id`.

        The internal key ID and secret key value itself are not returned in the
        response.

        :param user_id: the user ID
        :param key_id: the ID of the API key to get
        :return: the API key
        """

        document = await self._mongo.keys.find_one(
            {"id": key_id, "user.id": user_id},
            {
                "_id": False,
                "user": False,
            },
        )

        user = await self.data.users.get(user_id)

        if not document:
            raise ResourceNotFoundError

        document = await apply_transforms(
            base_processor(document),
            [
                AttachGroupsTransform(self._pg),
            ],
        )

        document["permissions"] = limit_permissions(
            document["permissions"], user.permissions.dict()
        )

        return APIKey(
            **{
                **document,
                "groups": sorted(document["groups"], key=lambda g: g["name"]),
            }
        )

    async def create_key(
        self,
        data: CreateKeysRequest,
        user_id: str,
    ) -> tuple[str, APIKey]:
        """
        Create a new API key.

        :param data: the fields to create a new API key
        :param user_id: the user ID
        :return: the API key
        """

        user = await self.data.users.get(user_id)

        raw, hashed = virtool.utils.generate_key()

        id_ = await virtool.account.mongo.get_alternate_id(self._mongo, data.name)

        await self._mongo.keys.insert_one(
            {
                "_id": hashed,
                "id": id_,
                "created_at": virtool.utils.timestamp(),
                "groups": [group.id for group in user.groups],
                "name": data.name,
                "permissions": data.permissions.dict(exclude_unset=True),
                "user": {"id": user_id},
            }
        )

        return raw, await self.get_key(user_id, id_)

    async def delete_keys(self, user_id: str):
        """
        Delete all API keys for the account associated with the requesting session.

        :param user_id: the user ID
        """
        await self._mongo.keys.delete_many({"user.id": user_id})

    async def update_key(
        self, user_id: str, key_id: str, data: UpdateKeyRequest
    ) -> APIKey:
        """
        Change the permissions for an existing API key.

        :param user_id: the user ID
        :param key_id: the ID of the API key to update
        :param data: permissions update
        :return: the API key
        """
        if data.permissions is None:
            update = {}
        else:
            update = data.permissions.dict(exclude_unset=True)

        if not await self._mongo.keys.count_documents({"id": key_id}):
            raise ResourceNotFoundError()

        new_permissions = {
            **(
                await get_one_field(
                    self._mongo.keys, "permissions", {"id": key_id, "user.id": user_id}
                )
            ),
            **update,
        }

        await self._mongo.keys.update_one(
            {"id": key_id},
            {"$set": {"permissions": new_permissions}},
        )

        return await self.get_key(user_id, key_id)

    async def delete_key(self, user_id: str, key_id: str):
        """
        Delete an API key by its id.

        :param user_id: the user ID
        :param key_id: the ID of the API key to delete
        """
        delete_result = await self._mongo.keys.delete_one(
            {"id": key_id, "user.id": user_id}
        )

        if delete_result.deleted_count == 0:
            raise ResourceNotFoundError()

    async def login(self, data: CreateLoginRequest) -> str:
        """
        Create a new session for the user with `username`.

        :param data: the login data
        :return: string representation of user_id
        """
        # When `remember` is set, the session will last for 1 month instead of the
        # 1-hour default

        # Re-render the login page with an error message if the username doesn't
        # correlate to a user_id value in
        # the database and/or password are invalid.
        document = await self._mongo.users.find_one({"handle": data.username})

        if not document or not await validate_credentials(
            self._mongo, document["_id"], data.password
        ):
            raise ResourceError()

        return document["_id"]

    async def get_reset_session(
        self, ip: str, user_id: str, session_id: str, remember: bool
    ) -> tuple[Session, str]:
        """
        Check if user password should be reset and return a reset code if it
        should be.

        :param ip: the ip address of the requesting client
        :param user_id: the login session ID
        :param session_id: the id of the session getting the reset code
        :param remember: boolean indicating whether the sessions should be remembered
        """

        if await get_one_field(self._mongo.users, "force_reset", user_id):
            await self.data.sessions.delete(session_id)
            return await self.data.sessions.create_reset(ip, user_id, remember)

        raise ResourceError

    async def logout(self, old_session_id: str, ip: str) -> Session:
        """
        Invalidates the requesting session, effectively logging out the user.

        :param old_session_id: the ID of the old session
        :param ip: the ip address of the client
        :return: the session_id, session, and session token
        """
        await self.data.sessions.delete(old_session_id)
        return await self.data.sessions.create_anonymous(ip)

    async def reset(
        self, session_id: str, data: ResetPasswordRequest, ip: str
    ) -> tuple[Session, str]:
        """
        Resets the password for a session user.

        :param session_id: the ID of the session to reset
        :param data: the data needed to reset session
        :param ip: the ip address of the client
        """

        session = await self.data.sessions.get_reset(session_id, data.reset_code)

        await self.data.sessions.delete(session_id)

        await self.data.users.update(
            session.reset.user_id,
            UpdateUserRequest(force_reset=False, password=data.password),
        )

        return await self.data.sessions.create_authenticated(
            ip, session.reset.user_id, remember=session.reset.remember
        )
