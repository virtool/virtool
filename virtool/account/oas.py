from typing import Union, Optional

from pydantic import BaseModel, constr, Field, root_validator, validator
from virtool_core.models.account import Account, AccountSettings, check_email, APIKey
from virtool_core.models.enums import QuickAnalyzeWorkflow
from virtool_core.models.validators import prevent_none

from virtool.groups.oas import PermissionsUpdate


class UpdateAccountRequest(BaseModel):
    """
    Fields for updating a user account.
    """

    email: constr(strip_whitespace=True) | None = Field(description="an email address")
    old_password: str | None = Field(description="the old password for verification")
    password: str | None = Field(description="the new password")

    class Config:
        schema_extra = {
            "example": {
                "email": "dev@virtool.ca",
                "password": "foo_bar_1",
                "old_password": "hello_world",
            }
        }

    @root_validator
    def check_password(cls, values: Union[str, constr]):
        """
        Checks if old_password has also been input if a new password
        is provided.
        """
        old_password, password = values.get("old_password"), values.get("password")

        if password:
            if not old_password:
                raise ValueError(
                    "The old password needs to be given in order for the password to be changed"
                )
        else:
            if old_password:
                raise ValueError(
                    "The new password needs to be given in order for the password to be changed"
                )

        return values

    _email_validation = validator("email", allow_reuse=True)(check_email)

    _prevent_none = prevent_none("*")


class UpdateAccountResponse(Account):
    class Config:
        schema_extra = {
            "example": {
                "administrator_role": None,
                "email": "dev@virtool.ca",
                "groups": [],
                "handle": "bob",
                "id": "test",
                "last_password_change": "2015-10-06T20:00:00Z",
                "permissions": {
                    "cancel_job": False,
                    "create_ref": False,
                    "create_sample": False,
                    "modify_hmm": False,
                    "modify_subtraction": False,
                    "remove_file": False,
                    "remove_job": False,
                    "upload_file": False,
                },
                "primary_group": {"id": 6, "name": "Technicians"},
                "settings": {
                    "quick_analyze_workflow": "pathoscope_bowtie",
                    "show_ids": True,
                    "show_versions": True,
                    "skip_quick_analyze_dialog": True,
                },
            }
        }


class UpdateSettingsRequest(BaseModel):
    """
    Fields for updating a user account's settings.
    """

    show_ids: Optional[bool] = Field(
        description="show document ids in client where possible"
    )
    skip_quick_analyze_dialog: Optional[bool] = Field(
        description="donâ€™t show the quick analysis dialog"
    )
    quick_analyze_workflow: Optional[QuickAnalyzeWorkflow] = Field(
        description="workflow to use for quick analysis"
    )
    show_versions: Optional[bool] = Field(
        description="show document versions in client where possible"
    )

    class Config:
        schema_extra = {"example": {"show_ids": False}}

    _prevent_none = prevent_none("*")


class CreateKeysRequest(BaseModel):
    name: constr(strip_whitespace=True, min_length=1) = Field(
        description="a non-unique name for the API key"
    )
    permissions: Optional[PermissionsUpdate] = Field(
        default=PermissionsUpdate(),
        description="an object describing the permissions the new key will have. "
        "Any unset permissions will default to false",
    )

    class Config:
        schema_extra = {
            "example": {"name": "Foobar", "permissions": {"create_sample": True}}
        }

    _prevent_none = prevent_none("permissions")


class CreateAPIKeyResponse(APIKey):
    key: str

    class Config:
        schema_extra = {
            "example": {
                "created_at": "2015-10-06T20:00:00Z",
                "groups": [],
                "id": "foobar_0",
                "key": "raw_key",
                "name": "Foobar",
                "permissions": {
                    "cancel_job": False,
                    "create_ref": False,
                    "create_sample": True,
                    "modify_hmm": False,
                    "modify_subtraction": False,
                    "remove_file": False,
                    "remove_job": False,
                    "upload_file": False,
                },
            }
        }


class UpdateKeyRequest(BaseModel):
    permissions: Optional[PermissionsUpdate] = Field(
        description="a permission update comprising an object keyed by permissions "
        "with boolean values"
    )

    class Config:
        schema_extra = {"example": {"permissions": {"modify_subtraction": True}}}

    _prevent_none = prevent_none("permissions")


class APIKeyResponse(APIKey):
    class Config:
        schema_extra = {
            "example": {
                "created_at": "2015-10-06T20:00:00Z",
                "groups": [],
                "id": "foobar_0",
                "name": "Foobar",
                "permissions": {
                    "cancel_job": False,
                    "create_ref": False,
                    "create_sample": True,
                    "modify_hmm": False,
                    "modify_subtraction": False,
                    "remove_file": False,
                    "remove_job": False,
                    "upload_file": False,
                },
            }
        }


class CreateLoginRequest(BaseModel):
    username: constr(min_length=1) = Field(description="account username")
    password: constr(min_length=1) = Field(description="account password")
    remember: Optional[bool] = Field(
        default=False,
        description="value determining whether the session will last for 1 month or "
        "1 hour",
    )

    class Config:
        schema_extra = {
            "example": {
                "username": "foobar",
                "password": "p@ssword123",
                "remember": False,
            }
        }

    _prevent_none = prevent_none("remember")


class LoginResponse(BaseModel):
    class Config:
        schema_extra = {"example": {"reset": False}}


class ResetPasswordRequest(BaseModel):
    password: str
    reset_code: str

    class Config:
        schema_extra = {
            "example": {
                "password": "p@ssword123",
                "reset_code": "4bcda8b3bcaf5f84cc6e26a3d23a6179f29d356e43c9ced1b6de0d8f4946555e",
            }
        }


class AccountResetPasswordResponse(BaseModel):
    class Config:
        schema_extra = {"example": {"login": False, "reset": False}}


class AccountResponse(Account):
    class Config:
        schema_extra = {
            "example": {
                "administrator_role": None,
                "groups": [],
                "handle": "bob",
                "id": "test",
                "last_password_change": "2015-10-06T20:00:00Z",
                "permissions": {
                    "cancel_job": False,
                    "create_ref": False,
                    "create_sample": False,
                    "modify_hmm": False,
                    "modify_subtraction": False,
                    "remove_file": False,
                    "remove_job": False,
                    "upload_file": False,
                },
                "primary_group": {
                    "id": 5,
                    "name": "Technician",
                },
                "settings": {
                    "quick_analyze_workflow": "pathoscope_bowtie",
                    "show_ids": True,
                    "show_versions": True,
                    "skip_quick_analyze_dialog": True,
                },
            }
        }


class AccountSettingsResponse(AccountSettings):
    class Config:
        schema_extra = {
            "example": {
                "skip_quick_analyze_dialog": True,
                "show_ids": True,
                "show_versions": True,
                "quick_analyze_workflow": "pathoscope_bowtie",
            }
        }


class ListAPIKeysResponse(APIKey):
    class Config:
        schema_extra = {
            "example": [
                {
                    "created_at": "2015-10-06T20:00:00Z",
                    "groups": [],
                    "id": "baz_1",
                    "name": "Baz",
                    "permissions": {
                        "cancel_job": False,
                        "create_ref": False,
                        "create_sample": True,
                        "modify_hmm": False,
                        "modify_subtraction": False,
                        "remove_file": False,
                        "remove_job": False,
                        "upload_file": False,
                    },
                }
            ]
        }
