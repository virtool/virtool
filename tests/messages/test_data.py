import pytest
from sqlalchemy.ext.asyncio import AsyncSession, AsyncEngine

from virtool.data.errors import ResourceConflictError
from virtool.messages.data import MessagesData
from virtool.messages.models import SQLInstanceMessage
from virtool.messages.oas import CreateMessageRequest, UpdateMessageRequest
from virtool.mongo.core import DB


@pytest.fixture
async def messages_data(pg: AsyncEngine, mongo: DB) -> MessagesData:
    return MessagesData(pg, mongo)


class TestGet:
    async def test_active(self, snapshot, static_time, messages_data, pg, fake2):

        user = await fake2.users.create()

        async with AsyncSession(pg) as session:
            session.add_all(
                [
                    SQLInstanceMessage(
                        id=1,
                        color="blue",
                        message="This is test message 1",
                        created_at=static_time.datetime,
                        updated_at=static_time.datetime,
                        user=user.id,
                    ),
                    SQLInstanceMessage(
                        id=2,
                        color="yellow",
                        message="This is test message 2",
                        created_at=static_time.datetime,
                        updated_at=static_time.datetime,
                        user=user.id
                    )
                ]
            )
            await session.commit()

        assert await messages_data.get() == snapshot

    async def test_inactive(self, static_time, messages_data, pg):

        async with AsyncSession(pg) as session:
            session.add(
                SQLInstanceMessage(
                    id=1,
                    active=False,
                    color="grey",
                    message="This is a test message",
                    created_at=static_time.datetime,
                    updated_at=static_time.datetime,
                    user="test",
                )
            )
            await session.commit()

        with pytest.raises(ResourceConflictError):
            await messages_data.get()


async def test_create(snapshot, static_time, messages_data, fake2):

    user = await fake2.users.create()

    create_request = CreateMessageRequest(color="blue", message="This is a test message")

    await messages_data.create(create_request, user.id)

    assert await messages_data.get() == snapshot


async def test_update(snapshot, pg, static_time, messages_data, fake2):

    user = await fake2.users.create()

    async with AsyncSession(pg) as session:
        session.add(
            SQLInstanceMessage(
                id=1,
                color="yellow",
                message="This is a test message",
                created_at=static_time.datetime,
                updated_at=static_time.datetime,
                user=user.id,
            )
        )
        await session.commit()

    update_request = UpdateMessageRequest(color="red", message="Updated Message")

    await messages_data.update(update_request)

    assert await messages_data.get() == snapshot
