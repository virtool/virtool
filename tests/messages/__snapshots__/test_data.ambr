# serializer version: 1
# name: TestGet.test_active
  dict({
    'active': True,
    'color': <MessageColor.yellow: 'yellow'>,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'id': 2,
    'message': 'This is test message 2',
    'updated_at': datetime.datetime(2015, 10, 6, 20, 0),
    'user': dict({
      'handle': 'leeashley',
      'id': 'bf1b993c',
    }),
  })
# ---
# name: test_create
  dict({
    'active': True,
    'color': <MessageColor.blue: 'blue'>,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'id': 1,
    'message': 'This is a test message',
    'updated_at': datetime.datetime(2015, 10, 6, 20, 0),
    'user': dict({
      'handle': 'leeashley',
      'id': 'bf1b993c',
    }),
  })
# ---
# name: test_update
  dict({
    'active': True,
    'color': <MessageColor.red: 'red'>,
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'id': 1,
    'message': 'Updated Message',
    'updated_at': datetime.datetime(2015, 10, 6, 20, 0),
    'user': dict({
      'handle': 'leeashley',
      'id': 'bf1b993c',
    }),
  })
# ---
